src/common/system_executor.cpp

```diff
@@ -0,0 +1,237 @@
+
+/* We use an atomic void(*)() rather than a std::function to avoid needing to take a mutex
+ * in exception context, as taking a mutex can itself throw an exception!
+ */
+std::atomic<void(*)()> exception_handler = &just_rethrow_exception;
+
+class Worker
+{
+public:
+    Worker()
+        : work{[](){}},
+          work_available{0}
+    {
+        std::promise<std::atomic<bool>*> shutdown_channel;
+        auto resolved_shutdown_location = shutdown_channel.get_future();
+
+        thread = std::thread{
+            &work_loop,
+            this,
+            std::move(shutdown_channel)};
+
+        shutdown = resolved_shutdown_location.get();
+    }
+
+    ~Worker() noexcept
+    {
+        if (thread.joinable())
+        {
+            signal_work_thread_to_halt();
+            if (thread.get_id() == std::this_thread::get_id())
+            {
+                /* We're being destroyed from our own work thread.
+                 * This can only happen when destroying the work functor at me->work = [](){}
+                 * in work_loop results in destroying the last shared pointer to *this.
+                 *
+                 * At this point, work_loop will exit without touching anything
+                 */
+                thread.detach();
+            }
+            else
+            {
+                thread.join();
+            }
+        }
+    }
+
+    void set_work(std::function<void()>&& to_do)
+    {
+        // Precondition: this Worker is idle (work_available is unraised, work = [](){})
+        work = std::move(to_do);
+        work_available.release();
+    }
+
+    /**
+     * Halt the worker, dropping any work scheduled but not yet executing, and wait for completion.
+     */
+    void stop()
+    {
+        signal_work_thread_to_halt();
+        thread.join();
+    }
+private:
+    void signal_work_thread_to_halt()
+    {
+        *shutdown = true;
+        /* We need to release the workloop thread by raising the signal.
+         * However, it's UB to raise the semaphore if it's already raised.
+         * So, first, consume any pending signal…
+         */
+        work_available.try_acquire();
+        /* work_available is now guaranteed not-signalled, so it's OK to
+         * unconditionally raise the signal and so release the workloop thread.
+         */
+        work_available.release();
+    }
+
+    static void work_loop(Worker* me, std::promise<std::atomic<bool>*>&& shutdown_channel)
+    {
+        mir::set_thread_name("Mir/Workqueue");
+
+        std::atomic<bool> shutdown{false};
+        shutdown_channel.set_value(&shutdown);
+
+        while (!shutdown)
+        {
+            me->work_available.acquire();
+            auto work = std::move(me->work);
+            me->work = [](){};
+            work();
+        }
+    }
+
+    std::function<void()> work;
+    std::atomic<bool>* shutdown;
+    std::binary_semaphore work_available;
+    std::thread thread;
+};
+
+class ThreadPool : public mir::SystemExecutor
```

Comment:
Why doesn't this use `mir::thread::BasicThreadPool` (the thread pool used by `MultiThreadedCompositor`)? Their implementations look quite a bit different, but do they have different requirements?

Reply:
So `BasicThreadPool` lives in `mirserver`, and is tied to symbols there, so it can't go into `mircommon` without changes (which, admittedly, would be pretty easy). It also has a bunch of superfluous functionality that doesn't get used at all - particularly, scheduling work on a specific thread and each thread having a queue of work. It is also not self-limiting - a manual call to `shrink()` is necessary in order to reap threads (although, again, this could be worked around).

It'd be simple to get `MultiThreadedCompositor` to use `mir::system_executor` instead; I'll try that.